src/printer.js

```diff
@@ -50,26 +48,38 @@ const {
  * @param {(path: import('prettier').AstPath<any>) => import('prettier').Doc} print
  */
 function printTopLevelParts(node, path, opts, print) {
-  const parts = {
-    frontmatter: [],
-    markup: [],
-    styles: [],
-  };
+  let docs = [];
 
+  const normalize = (doc) => [stripTrailingHardline(doc), softline];
+
+  // frontmatter always comes first
   if (node.module) {
-    parts.frontmatter.push(path.call(print, 'module'));
+    const subDoc = normalize(path.call(print, 'module'));
+    docs.push(subDoc);
   }
 
-  if (node.css && node.css.length) {
-    parts.styles.push(path.call(print, 'css'));
+  // markup and styles follow, whichever the user prefers (default: markup, styles)
+  for (const section of parseSortOrder(opts.astroSortOrder)) {
+    switch (section) {
+      case 'markup': {
+        const subDoc = path.call(print, 'html');
+        if (!isEmptyDoc(subDoc)) docs.push(normalize(subDoc));
+        break;
+      }
+      case 'styles': {
+        const subDoc = path.call(print, 'css');
+        if (!isEmptyDoc(subDoc)) docs.push(normalize(subDoc));
+        break;
+      }
+    }
   }
 
-  if (node.html && !isObjEmpty(node.html)) {
-    parts.markup.push(path.call(print, 'html'));
-  }
+  // remove trailing softline, if any
+  const lastDoc = docs[docs.length - 1];
+  const lastItem = lastDoc[lastDoc.length - 1];
+  if (lastItem.type === 'line') docs[docs.length - 1].pop();
```

Comment:
Lastly, this is a matter of preference, but I just took out the final empty `softline` that causes an empty final line.

I like final newlines in documents, but for our Prettier parser, if we’re calling this from somewhere else, I think it’s simpler if it doesn’t have a final newline. But either way I think this is arbitrary.

Reply:
Well, it should output a newline at the end b/c of [files always end with a new line](https://prettier.io/docs/en/rationale.html#empty-lines)

> Empty lines

> It turns out that empty lines are very hard to automatically generate. The approach that Prettier takes is to preserve empty lines the way they were in the original source code. There are two additional rules:

>    Prettier collapses multiple blank lines into a single blank line.
    Empty lines at the start and end of blocks (and whole files) are removed. (Files always end with a single newline, though.)

Reply:
Ah so now I remember why I did this. When we parse `.astro` within Markdown that counts as a separate document. So we get extra newlines within markdown blocks:

```diff
  ```astro
  <!-- Example1.astro - Static HTML is a valid Astro component! -->
  <div class="example-1">
    <h1>Hello world!</h1>
  </div>
+
```

I’m not sure how to avoid inserting this newline everywhere because it seems to happen at this top level where I removed it. Maybe this can be a followup PR?

I agree that files should end with new lines but I’m also wondering if that’s Prettier’s responsibility if it doesn’t know what’s the top-level “file” and what‘s an embedded document.

Reply:
I tried changing line 53 to 
```js
const normalize = (doc) => [stripTrailingHardline(doc), hardline];
```
and it seems to be working.

Reply:
> Ah so now I remember why I did this. When we parse `.astro` within Markdown that counts as a separate document. So we get extra newlines within markdown blocks:
> 
> ```diff
>   ```astro
>   <!-- Example1.astro - Static HTML is a valid Astro component! -->
>   <div class="example-1">
>     <h1>Hello world!</h1>
>   </div>
> +
> ```
> 
> I’m not sure how to avoid inserting this newline everywhere because it seems to happen at this top level where I removed it. Maybe this can be a followup PR?
> 
> I agree that files should end with new lines but I’m also wondering if that’s Prettier’s responsibility if it doesn’t know what’s the top-level “file” and what‘s an embedded document.

That does seem complicated! I wonder whether this has something to do with the `embed` function? https://prettier.io/docs/en/plugins.html#optional-embed

Reply:
🤔 Now that I think about it, this may be something we’d have to strip out inside the triple backtick itself. Perhaps we could strip newlines from there only. I think that’d solve the markdown problem (potentially).